profit/sur/sur.py

```diff
@@ -29,31 +29,97 @@ class Surrogate(ABC):
         output_ndim (int): Dimension of output data.
-        multi_output (bool): If a multi-output fit is desired. If False, the excess output dimensions are used as
-            independent supporting points.
+        multi_output (bool): True, if more than one output variable is defined in the
+            config file. If False, excess output dimensions are used as independent supporting points.
+        encoder (list of profit.sur.encoders.Encoder): For now, inputs of kind 'LogUniform' are encoded with 'log10'
+            and all input and output data is normalized. Can be modified in the config file using the format
+            e.g. [['log10', [0], False], ['normalization', [0, 1], False]].
 
     Default parameters:
         surrogate: GPy
-        save: False
+        save: ./model_{surrogate_label}.hdf5
         load: False
-        multi_output: False
+        fixed_sigma_n: False
+        encoder: [['log10', [log_input_cols], False], ['normalization', [input_cols], False],
+            ['normalization', [output_cols], True]]
     """
 
     _surrogates = {}  # All surrogates are registered here
     _defaults = {'surrogate': 'GPy',  # Default surrogate configuration parameters
-                 'save': False, 'load': False,
-                 'multi_output': False}
+                 'save': './model.hdf5',
+                 'load': False,
+                 'fixed_sigma_n': False,
+                 'encoder': []}
 
     def __init__(self):
         self.trained = False
         self.fixed_sigma_n = False
-        self.multi_output = False
+        self.multi_output = False  # Inferred from output variables in config.
 
         self.Xtrain = None
         self.ytrain = None
-        self.ndim = None
+        self.ndim = None  # TODO: Consistency between len(base_config['input']) and self.Xtrain.shape[-1]
         self.output_ndim = 1
 
+        self.encoder = []
+
+    def encode_training_data(self):
+        """Encodes the input and output training data.
+        """
+        for enc in self.encoder:
+            if enc.output:
+                self.ytrain = enc.encode(self.ytrain)
+            else:
+                self.Xtrain = enc.encode(self.Xtrain)
+
+    def decode_training_data(self):
+        """Applies the decoding function of the encoder in reverse order on the input and output training data.
+        """
+        for enc in self.encoder[::-1]:
+            if enc.output:
+                self.ytrain = enc.decode(self.ytrain)
+            else:
+                self.Xtrain = enc.decode(self.Xtrain)
+
+    def encode_predict_data(self, x):
+        """Transforms the input prediction points according to the encoder used for training.
+
+        Parameters:
+            x (ndarray): Prediction input points.
+
+        Returns:
+            ndarray: Encoded and normalized prediction points.
+        """
+        for enc in self.encoder:
+            if not enc.output:
+                if enc.label == 'normalization':
```

Comment:
I guess this will be changed to nice decoupled object-oriented design later, right?

Reply:
Exactly, the idea is to set up a modular `Encoder` class just like the surrogates, where also custom encoder functions can be registered. This will be done in v0.5.